File: web/src/figures/fig_1_10.ts
import Plotly from 'plotly.js-dist-min';
import { Rnl, radialDistribution, constants } from '../common/hydrogenic';
import { slider } from '../common/ui';

const controls = document.getElementById('controls')!;
const plot = document.getElementById('plot')!;
const caption = document.getElementById('caption')!;
caption.textContent = 'FIGURE 1.10 s‑orbitals showing radial nodes (roots of R_n0).';

const nSlider = slider(controls, 'n', 1, 4, 3);
const zSlider = slider(controls, 'Z', 1, 3, 1);

const findNodes = (n: number, Z: number) => {
  const nodes: number[] = [];
  let prev = Rnl(n, 0, 1e-6, Z);
  for (let i = 1; i <= 800; i++) {
    const r = (i / 800) * (20 * constants.a0);
    const v = Rnl(n, 0, r, Z);
    if (prev === 0) prev = v;
    if (prev * v < 0) nodes.push(r);
    prev = v;
  }
  return nodes;
};

const compute = () => {
  const n = parseInt(nSlider.value, 10);
  const Z = parseInt(zSlider.value, 10);
  const xs: number[] = [];
  const P: number[] = [];
  for (let i = 0; i <= 400; i++) {
    const r = (20 * constants.a0 * i) / 400;
    xs.push((Z * r) / constants.a0);
    P.push(radialDistribution(n, 0, r, Z));
  }

  const nodes = findNodes(n, Z).map((r) => (Z * r) / constants.a0);

  const traces: any[] = [
    { x: xs, y: P, mode: 'lines', name: `P(r) ${n}s`, line: { color: '#5a4ea3' } }
  ];
  nodes.forEach((x, i) => {
    traces.push({ x: [x, x], y: [0, Math.max(...P)], mode: 'lines', name: `node ${i+1}`, line: { dash: 'dot', color: '#e67e22' }, showlegend: false });
  });

  Plotly.newPlot(plot, traces, {
    margin: { l: 60, r: 20, t: 20, b: 40 },
    xaxis: { title: 'x = Zr/a₀' },
    yaxis: { title: 'P(r)' },
    legend: { orientation: 'h' }
  }, { displayModeBar: false });
};

[nSlider, zSlider].forEach((el) => el.addEventListener('input', compute));
compute();

File: pluto/fig_1_10.jl
### A Pluto.jl notebook ###
# v0.19.0

using PlutoUI
using WGLMakie
include("common/hydrogenic.jl")
using .Hydrogenic

md"""# Figure 1.10
Radial nodes in s orbitals (roots of R_n0)."""

@bind n Slider(1:4, default=3)
@bind Z Slider(1:3, default=1)

xs = range(0, 30; length=500)
rs = [x*a0/Z for x in xs]
P = [radial_distribution(n,0,r; Z=Z) for r in rs]

nodes = Float64[]
prev = R_nl(n,0,rs[1]; Z=Z)
for r in rs[2:end]
    v = R_nl(n,0,r; Z=Z)
    if prev*v < 0
        push!(nodes, r)
    end
    prev = v
end

fig = Figure(resolution=(800,400))
ax = Axis(fig[1,1], xlabel="x = Zr/a0", ylabel="P(r)")
lines!(ax, xs, P, color=:purple)
for r in nodes
    v = Z*r/a0
    vline!(ax, v, color=:orange, linestyle=:dash)
end
fig
